### PR TITLE
Add additionalProperties feature

### PR DESCRIPTION
**Pull request by manukura**
Added additionalProperties for @Attributes and insert into schema when false (default is true)

### CODE REVIEW

**Comment by manukura** [comment]
This will insert the flag additionalProperties into the schema for a complex type when the @Attributes annotation is marked with additionalProperties false.

The default value is true, and there will be no change to the resulting schema unless it is set to false.

File: src/main/java/com/github/reinert/jjschema/v1/CustomSchemaWrapper.java [hunk 1]
```diff
@@ -285,6 +285,9 @@ protected void processAttributes(ObjectNode node, Class<?> type) {
             if (attributes.readonly()) {
             	node.put("readonly", attributes.readonly());
             }
+            if (! attributes.additionalProperties ()) {
```
Comment (reinert) [inline src/main/java/com/github/reinert/jjschema/v1/CustomSchemaWrapper.java]: No spaces after `!` and before `()`

File: src/main/java/com/github/reinert/jjschema/v1/CustomSchemaWrapper.java [hunk 2]
```diff
@@ -285,6 +285,9 @@ protected void processAttributes(ObjectNode node, Class<?> type) {
...
+                node.put ("additionalProperties", false);
```
Comment (reinert) [inline src/main/java/com/github/reinert/jjschema/v1/CustomSchemaWrapper.java]: no spaces between method name and parentheses

File: src/test/java/com/github/reinert/jjschema/v1/AdditionalPropertiesTest.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2014, Danilo Reinert (daniloreinert@growbit.com)
```
Comment (reinert) [inline src/test/java/com/github/reinert/jjschema/v1/AdditionalPropertiesTest.java]: 2017

File: src/main/java/com/github/reinert/jjschema/Attributes.java [hunk 4]
```diff
@@ -68,4 +68,6 @@
     long maxLength() default -1L;
     
     boolean readonly() default false;
+
+    boolean additionalProperties () default true;
```
Comment (reinert) [inline src/main/java/com/github/reinert/jjschema/Attributes.java]: no spaces between method name and parentheses

File: src/test/java/com/github/reinert/jjschema/v1/AdditionalPropertiesTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+    public void testGenerateSaleSchema() throws JsonProcessingException {
+        JsonNode schema = v4generator.createSchema(Sale.class);
+        // System.out.println(MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(schema));
+        assertFalse (schema.get ("additionalProperties").asBoolean ());
+        JsonNode properties = schema.get("properties");
+        assertEquals(2, properties.size());
+
+        // ensure additional properties is in the nested schema
+        assertFalse (properties.findValue ("additionalProperties").asBoolean ());
```
Comment (reinert) [inline src/test/java/com/github/reinert/jjschema/v1/AdditionalPropertiesTest.java]: no spaces between method names and parentheses

File: src/test/java/com/github/reinert/jjschema/v1/AdditionalPropertiesTest.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        // System.out.println(MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(schema));
+        JsonNode properties = schema.get("properties");
+        assertEquals(2, properties.size());
+        assertFalse (schema.get ("additionalProperties").asBoolean ());
+    }
+
+    public void testGenerateSaleSchema() throws JsonProcessingException {
+        JsonNode schema = v4generator.createSchema(Sale.class);
+        // System.out.println(MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(schema));
+        assertFalse (schema.get ("additionalProperties").asBoolean ());
```
Comment (reinert) [inline src/test/java/com/github/reinert/jjschema/v1/AdditionalPropertiesTest.java]: no spaces between method names and parentheses

File: src/test/java/com/github/reinert/jjschema/v1/AdditionalPropertiesTest.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    public AdditionalPropertiesTest (String testName) {
+        super(testName);
+    }
+
+    public void testGenerateSaleItemSchema() throws JsonProcessingException {
+        JsonNode schema = v4generator.createSchema(SaleItem.class);
+        // System.out.println(MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(schema));
+        JsonNode properties = schema.get("properties");
+        assertEquals(2, properties.size());
+        assertFalse (schema.get ("additionalProperties").asBoolean ());
```
Comment (reinert) [inline src/test/java/com/github/reinert/jjschema/v1/AdditionalPropertiesTest.java]: no spaces between method names and parentheses

File: src/test/java/com/github/reinert/jjschema/v1/AdditionalPropertiesTest.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+/**
+ * @author manukura
+ */
+public class AdditionalPropertiesTest extends TestCase {
+
+    static ObjectMapper MAPPER = new ObjectMapper();
+    JsonSchemaFactory v4generator = new JsonSchemaV4Factory();
+
+    public AdditionalPropertiesTest (String testName) {
```
Comment (reinert) [inline src/test/java/com/github/reinert/jjschema/v1/AdditionalPropertiesTest.java]: no spaces between constructor name and parentheses

**Comment by manukura** [comment]
Sorry, I'm not sure what to do with this ... I have made the requested changes and checked them into my repo.

Does this automatically update the pull request, or do I have to raise another?

Also, it would be appreciated to know when the new snapshot is available, and especially when it would be targetted for a full release.

Thanks

**Comment by reinert** [comment]
Yeah, the PR is updated automatically. I'll make a minor release on Monday with your patch. Thanks.

**Comment by reinert** [comment]
I just published a new version with your patch.
In about an hour it will be availble in maven central.
Thank you for contributing.

**Comment by manukura** [comment]
Hi Danilo,

We are a bit out of date and are using Java 7.  Is there any way to have the release in java 7?

Thanks Frank

**Comment by reinert** [comment]
It would require me to split the project in a core module and java8 additional support module. This is not in my roadmap unfortunately.

**Comment by reinert** [comment]
The only place in the code that have Java8 specific elements is in https://github.com/reinert/JJSchema/blob/master/src/main/java/com/github/reinert/jjschema/SimpleTypeMappings.java#L62-L64

**Pull request merged**